test/e2e/framework/framework.go

```diff
@@ -285,7 +285,7 @@ func InitNumClusterNodes() error {
 
 func fetchClusterIDs() {
 	for i := range KubeClients {
-		gatewayNodes := findNodesByGatewayLabel(i, true)
+		gatewayNodes := findNodesByLabel(i, GatewayLabel, true)
```

Comment:
@tpantelis 

I faced another issue.
I have applied the below suggestion:
```
func (f *Framework) FindGatewayNodes(cluster ClusterIndex) []v1.Node {
	nodes, err := KubeClients[cluster].CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
		LabelSelector: labels.Set{GatewayLabel: "true"}.String(),
	})
	Expect(err).NotTo(HaveOccurred())

	return nodes.Items
}
```


The problem is that the `fetchClusterIDs` function that tries to call the `FindGatewayNodes` is yet initialized with the test framework.
So I can't really call it like `f.FindGatewayNodes(i)` or `framework.FindGatewayNodes(i)`.
Currently, it uses a regular function without the framework definition.

What do you think?

Reply:
@skitt 

The thing is that this function is widely used in the `submariner` repo tests, so I need to use the framework to defined it within the test.

Reply:
> 
> The thing is that this function is widely used in the `submariner` repo tests

`FindGatewayNodes`? That function is new with this PR.

Reply:
You're right.

The `FindGatewayNodes` and `FindNonGatewayNodes` functions replaces the following functions:
`func findNodesByGatewayLabel(cluster int, isGateway bool) []*v1.Node`
and
`func (f *Framework) FindNodesByGatewayLabel(cluster ClusterIndex, isGateway bool)`
which calls it.

In https://github.com/submariner-io/shipyard/blob/devel/test/e2e/framework/framework.go#L291, the call of the old function needs to be replaced with the new one.
`func (f *Framework) FindGatewayNodes(cluster ClusterIndex) []v1.Node`

But since it's a framework based function, I can't call it.

I can create a base function and call it with the new ones as it's done currently.

Reply:
`FindGatewayNodes` can be stand alone as @skitt mentioned:

`func FindGatewayNodes(cluster ClusterIndex) []v1.Node`

Reply:
@tpantelis @skitt 

Please, let me try to clarify the state of the issue.
Current patch proposes refactoring of the `findNodesByGatewayLabel` function.

Current implementation consists of two functions.
The first one is a base function:
`func findNodesByGatewayLabel(cluster int, isGateway bool) []*v1.Node` [link](https://github.com/submariner-io/shipyard/blob/devel/test/e2e/framework/nodes.go#L37)
This function is standalone and used within the `fetchClusterIDs` function [link](https://github.com/submariner-io/shipyard/blob/devel/test/e2e/framework/framework.go#L291).

The second, is a framework based function:
`func (f *Framework) FindNodesByGatewayLabel(cluster ClusterIndex, isGateway bool) []*v1.Node` [link](https://github.com/submariner-io/shipyard/blob/devel/test/e2e/framework/nodes.go#L33)
This function uses the base function described above with some input parameters.
This function used within the shipyard repo in the following place [link] (https://github.com/submariner-io/shipyard/blob/devel/test/e2e/framework/gateways.go#L191)

Beside the shipyard repo, the `FindNodesByGatewayLabel` function is used within the tests of `submariner` repository and called and test framework initialized function.
* https://github.com/submariner-io/submariner/blob/devel/test/e2e/redundancy/gateway_failover.go#L68
* https://github.com/submariner-io/submariner/blob/devel/test/e2e/redundancy/gateway_failover.go#L143
* https://github.com/submariner-io/submariner/blob/devel/test/e2e/redundancy/gateway_failover.go#L68
* https://github.com/submariner-io/submariner/blob/devel/test/e2e/redundancy/route_agent_restart.go#L50
* https://github.com/submariner-io/submariner/blob/devel/test/e2e/cluster/add_remove_cluster.go#L39
* https://github.com/submariner-io/submariner/blob/devel/test/e2e/cluster/add_remove_cluster.go#L61


With new changes, those calls needs to be replaced with the new functions from this patch, like:
```
func (f *Framework) FindGatewayNodes(cluster ClusterIndex) []v1.Node {
	nodes, err := KubeClients[cluster].CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
		LabelSelector: labels.Set{GatewayLabel: "true"}.String(),
	})
	Expect(err).NotTo(HaveOccurred())

	return nodes.Items
}
```

This function does not rely on any standalone function.
I can replace it for all the tests with no issues as test framework based function.
But I can't use it within the shipyard repo for `fetchClusterIDs` function, because the test framework is yet initialized at that place.
As I mentioned above, in current implementation it uses the standalone function, which the test framework function relies on.


I can convert this function into a standalone, but then it will require me to call this function from the tests in the following way:
Instead of `f.FindGatewayNodes(i)` use `framework.FindGatewayNodes(i)`.
I guess it should work.
But from my point of view it will create some sort of mess in the functions call flow.

Will appreciate you thoughts.
Thanks.

Reply:
`f.FindGatewayNodes(i)` and `framework.FindGatewayNodes(i)` are both just function calls. The difference is that `f.FindGatewayNodes(i)` calls the function associated with the type of `f` (and `f` can even be `nil`, as long as it’s typed), and the function can use `f` (but doesn’t, in this case); whereas `framework.FindGatewayNodes(i)` calls the function in the package aliased as `framework`. There’s no difference in terms of “function call flow”.

Reply:
Thank you.
I got the point.

Will check and verify that I'm able to call the function in the `framework.func_name` way.

Reply:
Set it as a standalone functions.
Working.
Thank you for the help.